source/YTEPs/YTEP-0037.rst

```diff
@@ -0,0 +1,261 @@
+
+From ``black``'s `documentation <https://black.readthedocs.io/en/stable/the_black_code_style.html>`_
+
+  The rules for horizontal whitespace can be summarized as: do whatever makes
+  pycodestyle happy.
+  (...)
+  The coding style used by Black can be viewed as a strict subset of PEP 8.
+
+so it is expected that ``black`` plays nicely with ``flake8`` by construction.
+``black`` applies an opinated style, and offers very little configuration options
+by design. Only the target line length can be changed. This makes it a critical 
+point, requiring discussion if this YTEP is approved.
+
+**maximal line length**
+
+The guidelines states that
+
+    Line widths should not be more than 80 characters
+
+Despite this being respected in most of the code base, there remains a large amount of
+outliers, that would be time-consuming to go through by hand. Taking the example of the
+yt-4.0 branch at the time of writing, there are 2158 lines exceeding 80 characters
+(~1.5% of the whole code base), or, visually
+
+.. image:: ../images/lw-ytep-37.png
+
+Note that when long strings are present, ``black`` will not attempt to split them to
+shorten the individual lines.
+This is most important in the case of dosctrings, and I explore the tools available to
+validate them hereafter (see `additional rules`_).
+
+There is a range of possible values we might give preference to. Python's standard
+library caps line-length at 79, pandas does so at 88. By default, ``black`` will target
+88, as its authors claim it reduces the total number of lines by some 10% (as compared
+to enforcing 80).
+
+In first drafting the PR linked above, I chose a line-length of 100, so as to minimize
+the amount of manual tweaking left to me after a ``black`` pass. I estimated that
+imposing a strict limit to 80 chars would leave 545 lines to be manually updated, while
+caping at 88 leaves a mere 135 (75% less work). As a reference, ``dask`` uses black's
+default settings, *and* allows flexibility for docstrings up to 120 characters through
+flake8.
+
+
+**sorting imports**
+
+PEP8 `recommends sorting imports statments <https://www.python.org/dev/peps/pep-0008/#imports>`_,
+Needless to say, the task is daunting and definitely not worthy of anyone's time if we
+had to go back and apply those rules manually to the code base.
+Luckilly, ``isort`` is able to check for and auto-apply those rules, so it can easily be
+added to the CI-linting process.
+
+Moreover, ``isort`` is configurable so that it allows for the definition of custom
+"sections" within import statements. This can be use to isolate imports from ``unyt``,
+which falls somewhere in between the default sections "third party" (external
+dependency) and "first party" (the project). This is done in 
+`#2592 <https://github.com/yt-project/yt/pull/2592>`_.
+
+To better highlight the way yt 4.0 depends on ``unyt``, I also propose that, within the
+code base, we import directly from ``unyt`` as often as possible, so as to limit
```

Comment:
Ok, so with `unyt` here do we want to create some specific guidelines beyond "as often as possible"? Where would we recommend importing from `yt.units` over `unyt` and vice-versa? Maybe @jzuhone has some opinions on this too.

Reply:
I don't see why we shouldn't always use `yt.units` instead? That way we're always consistent with our units.

Reply:
So my guidelines would be very simple (following what I've done in https://github.com/yt-project/yt/pull/2597):
- `YTArray` and `YTQuantity` can be imported from `yt.units`
- all the rest should be imported from `unyt`

Reply:
Hmmm, I'm not convinced about this yet and I think we should have more discussion about it. Why would importing from unyt directly be better than using yt.units? unyt's default units (mks) are in a different system than yt's (cgs). Won't we increase the probability of getting weird conflicts if we do imports of both?

Reply:
> Won't we increase the probability of getting weird conflicts if we do imports of both?

tbh I didn't think that what even possible. My mental model for `yt.units` is purely a wrapper module that we merely keep for convenience (avoid breaking downstream code if we don't need to). If this view is incorrect in any way (which seems to be the case with cgs VS mks systems), then it's expected that my recommandations don't make much sense. :/

Reply:
I should add that this view is based on the documentation in doc/source/yt4differences.rst, which may need to be slightly rephrased.

Reply:
Sorry just dropping in because I happened to read some notifications about this. 

Every Unit object had a unit registry object attached to it and every unit registry is associated with a unit system. The unit objects in the `unyt` namespace use a MKS unit system, the ones in `yt.units` use a CGS system. Try for yourself by doing eg `some_data.in_base()` using data crested from `yt.units` and `unyt` and you’ll get different results.

You can mix and match units with different unit systems, but the result of any operation inherits the unit system from the leftmost operand. This means that if internally in yt a unit ultimately came from the `unyt` namespace, a user might end up with a result that will ultimately be associated with an MKS unit system, perhaps causing confusion or even outright buggy behavior.

This means that inside yt and for users of yt the least risky approach is to always import things from `yt.units` or `ds.units` and to more or less ignore that unit objects can be imported from `unyt` too unless you are taking special care to make sure the unit systems and registries come out correct.

In principle yt could move to using the default unyt MKS unit system but that’s a different discussion. It’s a little annoying that we had to change this default when unyt got split off but me and John both agreed at the time that more people would think of MKS units as being a natural default than in yt’s astronomer-heavy userbase.

Reply:
Why do we want to keep `yt.units` at all? I thought the whole point of splitting out `unyt` was for `yt` to fully depend on it. It's fine to keep some sort of thin wrapper to provide backward compatibility, but shouldn't we "just" switch? Even if that means embracing MKS everywhere?

Reply:
Because we wanted to minimize back compat breaks for users who were already burned by the yt-2 to yt-3 transition.

Reply:
I think in 4.0, which I believe is the target here, most/all of `yt.units` now is a thin-layer.  And that means, among other things, that they have MKS base:

```
>>> yt.units.cm.base_value          
0.01
```

There are a handful of additional things still included, like `UnitContainer`, `display_ytarray`, some `create_code_unit_system` stuff and a default registry that includes `"h"` but for the most part it now imports directly from `unyt`.

Reply:
So `base_value` is the value in `unyt`'s internal unit system, which was indeed changed to an MKS unit system. However, the internal unit system is more of an implementation detail, and is not really exposed in the API beyond the `base_value` attribute Matt pointed out, unit objects each refer to a unit system distinct from the internal unit system that's used to generate default values for the results of operations (e.g. for unit simplication). The units in the yt namespace have a different unit system from the ones in the unyt namespace:

```
In [3]: yt.units.cm.registry.unit_system
Out[3]:
cgs Unit System
 Base Units:
  length: cm
  mass: g
  time: s
  temperature: K
  angle: rad
  luminous_intensity: cd
  logarithmic: Np
 Other Units:
  energy: erg
  specific_energy: erg/g
  pressure: dyn/cm**2
  force: dyn
  magnetic_field_cgs: G
  charge_cgs: statC
  current_cgs: statA
  power: erg/s

```
In [5]: unyt.cm.registry.unit_system
Out[5]:
mks Unit System
 Base Units:
  length: m
  mass: kg
  time: s
  temperature: K
  angle: rad
  current_mks: A
  luminous_intensity: cd
  logarithmic: Np
 Other Units:
  energy: J
  specific_energy: J/kg
  pressure: Pa
  force: N
  magnetic_field: T
  charge: C
  frequency: Hz
  power: W
  electric_potential: V
  capacitance: F
  inductance: H
  resistance: Ω
  magnetic_flux: Wb
  luminous_flux: lm
```

We did it this way to make it easier to keep backward compatibility in yt.

If you're interested in more detail about this in a slightly more abstract context, both the yt namespace and yt `Dataset` instances implement the pattern described in the unyt docs here: https://unyt.readthedocs.io/en/stable/usage.html#custom-unit-systems

Reply:
thanks @ngoldbaum for the clarifications !

> This means that if internally in yt a unit ultimately came from the unyt namespace, a user might end up with a result that will ultimately be associated with an MKS unit system, perhaps causing confusion or even outright buggy behavior.

> I think in 4.0, which I believe is the target here, most/all of yt.units now is a thin-layer. And that means, among other things, that they have MKS base
There are a handful of additional things still included, like UnitContainer, display_ytarray, some create_code_unit_system stuff and a default registry that includes "h" but for the most part it now imports directly from unyt.

In https://github.com/yt-project/yt/pull/2597 I only tried to avoid the additional import layer, I don't think any behaviour (internal or external) should be affected ?



> Why do we want to keep yt.units at all? I thought the whole point of splitting out unyt was for yt to fully depend on it. It's fine to keep some sort of thin wrapper to provide backward compatibility, but shouldn't we "just" switch? Even if that means embracing MKS everywhere?

> Because we wanted to minimize back compat breaks for users who were already burned by the yt-2 to yt-3 transition.

But is yt.units planned for deprecation at some point ? I don't know if we're the 3->4 transition will be comparable to 2->3 in terms of backward compatibility breaks, so maybe it's a good time to at least add deprecation warnings (or even get rid of the wrapper completely and just keep the yt-specific bits) ?

In any case, I feel like everyone here agrees it'd be best to avoid mixing imports from both unyt and yt.units
Please correct me if I'm mistaken but it seemed to me that the current state of yt.unit was meant to be transitory, and that yt.units would ultimately disappear in favour of unyt.